File: packages/dprint-plugin-typescript/src/utils/Stack.ts
import { throwError } from "./assertions";

export class Stack<T> {
    private readonly items: T[] = [];

    peek(): T | undefined {
        return this.items[this.items.length - 1];
    }

    push(item: T) {
        this.items.push(item);
    }

    popOrThrow() {
        const result = this.items.pop();

        if (result == null)
            return throwError("Tried to pop, but stack was empty. Maybe a pop was accidentally done elsewhere?");

        return result;
    }
}

File: packages/rust-dprint-plugin-typescript/src/utils/stack.rs
pub struct Stack<T> {
    items: Vec<T>,
}

impl<T> Stack<T> {
    pub fn new() -> Stack<T> {
        Stack { items: Vec::new() }
    }

    pub fn pop(&mut self) -> T {
        let result = self.items.pop();
        result.expect("Tried to pop, but the stack was empty. This indicates a bug where an item is being popped, but not pushed to the stack.")
    }

    pub fn push(&mut self, item: T) {
        self.items.push(item);
    }

    pub fn peek(&self) -> Option<&T> {
        self.items.last()
    }

    pub fn iter(&self) -> impl Iterator<Item = &T> {
        self.items.iter().rev()
    }
}
